charms/katib-db-manager/src/charm.py

```diff
@@ -136,17 +161,171 @@ def _check_leader(self):
+        """Check for existing database relations and process mysql relation if needed."""
+        # check for relational-db relation
+        # relying on KeyError to ensure that relational-db relation is not present
+        try:
+            relation = self.model.get_relation("relational-db")
+            if relation:
+                self.logger.warning(
+                    "Up-to-date database relation relational-db is already established."
+                )
+                self.logger.warning(f"{MYSQL_WARNING} {UNBLOCK_MESSAGE}")
+                self.model.unit.status = BlockedStatus(
+                    f"{MYSQL_WARNING} {UNBLOCK_MESSAGE} See logs"
+                )
+                return
+        except KeyError:
+            pass
+        # relational-db relation was not found, proceed with warnings
+        self.logger.warning(MYSQL_WARNING)
+        self.model.unit.status = MaintenanceStatus(f"Adding mysql relation. {MYSQL_WARNING}")
+        self._on_event(event)
+
+    def _on_mysql_relation_remove(self, event):
+        """Process removal of mysql relation."""
+        self.model.unit.status = MaintenanceStatus(f"Removing mysql relation. {MYSQL_WARNING}")
+        self._on_event(event)
+
+    def _on_relational_db_relation(self, event):
+        """Check for existing database relations and process relational-db relation if needed."""
+        # relying on KeyError to ensure that mysql relation is not present
+        try:
+            relation = self.model.get_relation("mysql")
+            if relation:
+                self.logger.warning(
+                    "Failed to create relational-db relation due to existing mysql relation."
+                )
+                self.logger.warning(f"{MYSQL_WARNING} {UNBLOCK_MESSAGE}")
+                self.model.unit.status = BlockedStatus(
+                    f"{MYSQL_WARNING} {UNBLOCK_MESSAGE} See logs"
+                )
+                return
+        except KeyError:
+            pass
+        # mysql relation was not found, proceed
+        self.model.unit.status = MaintenanceStatus("Adding relational-db relation")
+        self._on_event(event)
+
+    def _on_relational_db_relation_remove(self, event):
+        """Process removal of relational-db relation."""
+        self.model.unit.status = MaintenanceStatus("Removing relational-db relation")
+        self._on_event(event)
+
+    def _get_db_relation(self, relation_name):
+        """Retrieves relation with supplied relation name, if it is established.
+
+        Returns relation, if it is established, and raises error otherwise."""
+
+        try:
+            # retrieve relation data
+            relation = self.model.get_relation(relation_name)
+        except KeyError:
+            # relation was not found
+            relation = None
+        if not relation:
+            # relation is not established, raise an error
+            raise GenericCharmRuntimeError(f"Database relation {relation_name} is not established")
```

Comment:
or could it be that `self.model.get_relation(relation_name)` returns None ?
```suggestion
            raise GenericCharmRuntimeError(f"Database relation {relation_name} is not established")
```

Reply:
Actually I missed it here, it should also have `or empty`.
It was too complex to cover all error cases, so I think message indicating `not established or empty` should suffice.